File: src/migration/canonical-writer.ts
import path from 'node:path';
import { ensureDir, fileExists, fileHash, writeFile } from '../utils/files.js';
import type { FileRecord } from '../types.js';
import type { ParsedSetup } from './types.js';

export interface CanonicalWriterOptions {
  targetDir: string;
  parsedSetup: ParsedSetup;
  fileRecords: FileRecord[];
  dryRun?: boolean;
}

export interface CanonicalWriteResult {
  agents: string[];
  skills: string[];
  prompts: string[];
  rules: string[];
  rootConfig: string | null;
  skipped: string[];
}

function normalizeName(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '') || 'item';
}

function writeCanonicalFile(args: {
  destination: string;
  content: string;
  targetDir: string;
  source: string;
  fileRecords: FileRecord[];
  dryRun: boolean;
  skipped: string[];
}): boolean {
  if (fileExists(args.destination)) {
    args.skipped.push(path.relative(args.targetDir, args.destination));
    return false;
  }

  if (args.dryRun) {
    return true;
  }

  writeFile(args.destination, args.content);
  args.fileRecords.push({
    path: path.relative(args.targetDir, args.destination),
    hash: fileHash(args.destination),
    source: args.source,
  });

  return true;
}

export async function writeToCanonical(opts: CanonicalWriterOptions): Promise<CanonicalWriteResult> {
  const aiDir = path.join(opts.targetDir, '.ai');
  const dryRun = opts.dryRun === true;
  const result: CanonicalWriteResult = {
    agents: [],
    skills: [],
    prompts: [],
    rules: [],
    rootConfig: null,
    skipped: [],
  };

  if (!dryRun) {
    ensureDir(aiDir);
  }

  const agentsDir = path.join(aiDir, 'agents');
  if (!dryRun) ensureDir(agentsDir);
  for (const agent of opts.parsedSetup.agents ?? []) {
    const fileName = `${normalizeName(agent.id || agent.name)}.md`;
    const destination = path.join(agentsDir, fileName);
    const wrote = writeCanonicalFile({
      destination,
      content: agent.content,
      targetDir: opts.targetDir,
      source: `migrated:${agent.sourcePath}`,
      fileRecords: opts.fileRecords,
      dryRun,
      skipped: result.skipped,
    });
    if (wrote) {
      result.agents.push(path.relative(opts.targetDir, destination));
    }
  }

  const skillsDir = path.join(aiDir, 'skills');
  if (!dryRun) ensureDir(skillsDir);
  for (const command of opts.parsedSetup.commands ?? []) {
    const fileName = `${normalizeName(command.id || command.name)}.md`;
    const destination = path.join(skillsDir, fileName);
    const wrote = writeCanonicalFile({
      destination,
      content: command.content,
      targetDir: opts.targetDir,
      source: `migrated:${command.sourcePath}`,
      fileRecords: opts.fileRecords,
      dryRun,
      skipped: result.skipped,
    });
    if (wrote) {
      result.skills.push(path.relative(opts.targetDir, destination));
    }
  }

  const promptsDir = path.join(aiDir, 'prompts');
  if (!dryRun) ensureDir(promptsDir);
  for (const template of opts.parsedSetup.templates ?? []) {
    const fileName = `${normalizeName(template.id || template.name)}.md`;
    const destination = path.join(promptsDir, fileName);
    const wrote = writeCanonicalFile({
      destination,
      content: template.content,
      targetDir: opts.targetDir,
      source: `migrated:${template.sourcePath}`,
      fileRecords: opts.fileRecords,
      dryRun,
      skipped: result.skipped,
    });
    if (wrote) {
      result.prompts.push(path.relative(opts.targetDir, destination));
    }
  }

  const rulesDir = path.join(aiDir, 'rules');
  if (!dryRun) ensureDir(rulesDir);
  for (const rule of opts.parsedSetup.rules ?? []) {
    const fileName = `${normalizeName(rule.id || rule.category || 'rule')}.md`;
    const destination = path.join(rulesDir, fileName);
    const wrote = writeCanonicalFile({
      destination,
      content: rule.content,
      targetDir: opts.targetDir,
      source: `migrated:${rule.sourcePath}`,
      fileRecords: opts.fileRecords,
      dryRun,
      skipped: result.skipped,
    });
    if (wrote) {
      result.rules.push(path.relative(opts.targetDir, destination));
    }
  }

  if ((opts.parsedSetup.customSections?.length ?? 0) > 0) {
    const destination = path.join(aiDir, 'constitution', `${normalizeName(opts.parsedSetup.metadata?.adapter as string || opts.parsedSetup.projectName || 'migration')}.md`);
    const content = opts.parsedSetup.customSections
      .map((section) => `## ${section.title}\n\n${section.content}`)
      .join('\n\n');
    const wrote = writeCanonicalFile({
      destination,
      content,
      targetDir: opts.targetDir,
      source: `migrated:${opts.parsedSetup.metadata?.adapter ?? 'unknown'}`,
      fileRecords: opts.fileRecords,
      dryRun,
      skipped: result.skipped,
    });
    if (wrote) {
      result.rootConfig = path.relative(opts.targetDir, destination);
    }
  }

  return result;
}

File: internal/migration/canonicalwriter.go
package migration

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/ricardoborges-teachable/ai-setup/internal/files"
	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

// CanonicalWriteResult holds the result of writing files in canonical format.
type CanonicalWriteResult struct {
	Agents     []string
	Skills     []string
	Prompts    []string
	Rules      []string
	RootConfig string
	Skipped    []string
}

type canonicalTarget struct {
	TargetPath string
	SourcePath string
	Content    string
	Kind       ActionType
}

var nonAlphaNumRe = regexp.MustCompile(`[^a-z0-9]+`)
var leadingTrailingDashRe = regexp.MustCompile(`^-+|-+$`)

// WriteCanonical writes parsed data into the .ai/ canonical format.
// Ported from writeToCanonical in src/migration/canonical-writer.ts.
func WriteCanonical(targetDir string, setup *ParsedSetup, fileRecords *[]types.TrackedFile, dryRun bool) (*CanonicalWriteResult, error) {
	aiDir := filepath.Join(targetDir, ".ai")
	result := &CanonicalWriteResult{
		Agents:  []string{},
		Skills:  []string{},
		Prompts: []string{},
		Rules:   []string{},
		Skipped: []string{},
	}

	if !dryRun {
		if err := files.EnsureDir(aiDir); err != nil {
			return nil, err
		}
	}

	// Write agents.
	agentsDir := filepath.Join(aiDir, "agents")
	if !dryRun {
		if err := files.EnsureDir(agentsDir); err != nil {
			return nil, err
		}
	}
	for _, agent := range setup.Agents {
		fileName := normalizeName(agent.ID) + ".md"
		if fileName == ".md" {
			fileName = normalizeName(agent.Name) + ".md"
		}
		destination := filepath.Join(agentsDir, fileName)
		wrote, err := writeCanonicalFile(destination, []byte(agent.Content), targetDir, "migrated:"+agent.SourcePath, fileRecords, dryRun, result)
		if err != nil {
			return nil, fmt.Errorf("write %s: %w", destination, err)
		}
		if wrote {
			rel, _ := filepath.Rel(targetDir, destination)
			result.Agents = append(result.Agents, rel)
		}
	}

	// Write skills (commands).
	skillsDir := filepath.Join(aiDir, "skills")
	if !dryRun {
		if err := files.EnsureDir(skillsDir); err != nil {
			return nil, err
		}
	}
	for _, cmd := range setup.Commands {
		fileName := normalizeName(cmd.ID) + ".md"
		if fileName == ".md" {
			fileName = normalizeName(cmd.Name) + ".md"
		}
		destination := filepath.Join(skillsDir, fileName)
		wrote, err := writeCanonicalFile(destination, []byte(cmd.Content), targetDir, "migrated:"+cmd.SourcePath, fileRecords, dryRun, result)
		if err != nil {
			return nil, fmt.Errorf("write %s: %w", destination, err)
		}
		if wrote {
			rel, _ := filepath.Rel(targetDir, destination)
			result.Skills = append(result.Skills, rel)
		}
	}

	// Write prompts (templates).
	promptsDir := filepath.Join(aiDir, "prompts")
	if !dryRun {
		if err := files.EnsureDir(promptsDir); err != nil {
			return nil, err
		}
	}
	for _, tmpl := range setup.Templates {
		fileName := normalizeName(tmpl.ID) + ".md"
		if fileName == ".md" {
			fileName = normalizeName(tmpl.Name) + ".md"
		}
		destination := filepath.Join(promptsDir, fileName)
		wrote, err := writeCanonicalFile(destination, []byte(tmpl.Content), targetDir, "migrated:"+tmpl.SourcePath, fileRecords, dryRun, result)
		if err != nil {
			return nil, fmt.Errorf("write %s: %w", destination, err)
		}
		if wrote {
			rel, _ := filepath.Rel(targetDir, destination)
			result.Prompts = append(result.Prompts, rel)
		}
	}

	// Write rules.
	rulesDir := filepath.Join(aiDir, "rules")
	if !dryRun {
		if err := files.EnsureDir(rulesDir); err != nil {
			return nil, err
		}
	}
	for _, rule := range setup.Rules {
		fileName := normalizeName(rule.ID) + ".md"
		if fileName == ".md" {
			fileName = normalizeName(rule.Category) + ".md"
		}
		destination := filepath.Join(rulesDir, fileName)
		wrote, err := writeCanonicalFile(destination, []byte(rule.Content), targetDir, "migrated:"+rule.SourcePath, fileRecords, dryRun, result)
		if err != nil {
			return nil, fmt.Errorf("write %s: %w", destination, err)
		}
		if wrote {
			rel, _ := filepath.Rel(targetDir, destination)
			result.Rules = append(result.Rules, rel)
		}
	}

	// Write custom sections as root config.
	if len(setup.CustomSections) > 0 {
		adapterName := setup.Metadata["adapter"]
		if adapterName == "" {
			adapterName = setup.ProjectName
		}
		if adapterName == "" {
			adapterName = "migration"
		}
		destination := filepath.Join(aiDir, "constitution", normalizeName(adapterName)+".md")
		var sections []string
		for _, section := range setup.CustomSections {
			sections = append(sections, "## "+section.Title+"\n\n"+section.Content)
		}
		content := []byte(strings.Join(sections, "\n\n"))
		wrote, err := writeCanonicalFile(destination, content, targetDir, "migrated:"+adapterName, fileRecords, dryRun, result)
		if err != nil {
			return nil, fmt.Errorf("write %s: %w", destination, err)
		}
		if wrote {
			rel, _ := filepath.Rel(targetDir, destination)
			result.RootConfig = rel
		}
	}

	return result, nil
}

func PlannedCanonicalTargets(setup *ParsedSetup) []canonicalTarget {
	targets := []canonicalTarget{}

	for _, agent := range setup.Agents {
		targets = append(targets, canonicalTarget{
			TargetPath: filepath.Join(".ai", "agents", normalizeName(firstNonEmpty(agent.ID, agent.Name))+".md"),
			SourcePath: agent.SourcePath,
			Content:    agent.Content,
			Kind:       ActionTypeCreate,
		})
	}
	for _, command := range setup.Commands {
		targets = append(targets, canonicalTarget{
			TargetPath: filepath.Join(".ai", "skills", normalizeName(firstNonEmpty(command.ID, command.Name))+".md"),
			SourcePath: command.SourcePath,
			Content:    command.Content,
			Kind:       ActionTypeCreate,
		})
	}
	for _, template := range setup.Templates {
		targets = append(targets, canonicalTarget{
			TargetPath: filepath.Join(".ai", "prompts", normalizeName(firstNonEmpty(template.ID, template.Name))+".md"),
			SourcePath: template.SourcePath,
			Content:    template.Content,
			Kind:       ActionTypeCreate,
		})
	}
	for _, rule := range setup.Rules {
		targets = append(targets, canonicalTarget{
			TargetPath: filepath.Join(".ai", "rules", normalizeName(firstNonEmpty(rule.ID, rule.Category))+".md"),
			SourcePath: rule.SourcePath,
			Content:    rule.Content,
			Kind:       ActionTypeCreate,
		})
	}
	if len(setup.CustomSections) > 0 {
		adapterName := setup.Metadata["adapter"]
		if adapterName == "" {
			adapterName = setup.ProjectName
		}
		if adapterName == "" {
			adapterName = "migration"
		}
		targets = append(targets, canonicalTarget{
			TargetPath: filepath.Join(".ai", "constitution", normalizeName(adapterName)+".md"),
			SourcePath: setup.CustomSections[0].SourcePath,
			Content:    adapterName,
			Kind:       ActionTypeCreate,
		})
	}

	return targets
}

// writeCanonicalFile writes a file if it doesn't already exist. Returns true
// if the file was written (or would have been in dry-run mode).
func writeCanonicalFile(destination string, content []byte, targetDir, source string, fileRecords *[]types.TrackedFile, dryRun bool, result *CanonicalWriteResult) (bool, error) {
	if files.FileExists(destination) {
		rel, _ := filepath.Rel(targetDir, destination)
		result.Skipped = append(result.Skipped, rel)
		return false, nil
	}

	if dryRun {
		return true, nil
	}

	if err := files.WriteFile(destination, content, 0o644); err != nil {
		return false, err
	}

	hash, _ := files.FileHash(destination)
	rel, _ := filepath.Rel(targetDir, destination)
	*fileRecords = append(*fileRecords, types.TrackedFile{
		Path:   rel,
		Hash:   hash,
		Source: source,
		Owner:  types.FileOwnerMigrated,
		Status: types.FileStatusInstalled,
	})

	return true, nil
}

// normalizeName creates a safe filename from a free-form string.
func normalizeName(value string) string {
	s := strings.TrimSpace(value)
	s = strings.ToLower(s)
	s = nonAlphaNumRe.ReplaceAllString(s, "-")
	s = leadingTrailingDashRe.ReplaceAllString(s, "")
	if s == "" {
		return "item"
	}
	return s
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return "item"
}
